tsl/src/bgw_policy/job_api.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <funcapi.h>
+#include <miscadmin.h>
+#include <utils/builtins.h>
+
+#include <bgw/job.h>
+#include <bgw/job_stat.h>
+
+#include "job.h"
+#include "job_api.h"
+
+/* Default max runtime for a custom job is unlimited for now */
+#define DEFAULT_MAX_RUNTIME 0
+
+/* Right now, there is an infinite number of retries for custom jobs */
+#define DEFAULT_MAX_RETRIES -1
+/* Default retry period for reorder_jobs is currently 5 minutes */
+#define DEFAULT_RETRY_PERIOD 5 * USECS_PER_MINUTE
+
+#define ALTER_JOB_SCHEDULE_NUM_COLS 8
+
+/*
+ * CREATE FUNCTION add_job(
+ *   proc REGPROC,
+ *   schedule_interval INTERVAL,
+ *   config JSONB,
+ *   initial_start TIMESTAMPTZ,
+ *   scheduled BOOL DEFAULT true
+ * ) RETURNS INTEGER
+ */
+Datum
+job_add(PG_FUNCTION_ARGS)
+{
+	NameData application_name;
+	NameData custom_name;
+	NameData proc_name;
+	NameData proc_schema;
+	NameData owner_name;
+	Interval max_runtime = { .time = DEFAULT_MAX_RUNTIME };
+	Interval retry_period = { .time = DEFAULT_RETRY_PERIOD };
+	int32 job_id;
+
+	Oid owner = GetUserId();
+	Oid proc = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Interval *schedule_interval = PG_ARGISNULL(1) ? NULL : PG_GETARG_INTERVAL_P(1);
+	Jsonb *config = PG_ARGISNULL(2) ? NULL : PG_GETARG_JSONB_P(2);
+	bool scheduled = PG_ARGISNULL(4) ? true : PG_GETARG_BOOL(4);
+
+	/* Verify that the owner can create a background worker */
+	ts_bgw_job_validate_job_owner(owner, JOB_TYPE_CUSTOM);
+
+	/* Next, insert a new job into jobs table */
+	namestrcpy(&application_name, "Custom Job");
+	namestrcpy(&custom_name, "custom");
+	namestrcpy(&proc_schema, get_namespace_name(get_func_namespace(proc)));
```

Comment:
You should fail with a more useful error prior to this if proc is null or invalid.